wcfsetup/install/files/lib/form/NotificationUnsubscribeForm.class.php

```diff
@@ -0,0 +1,145 @@
+<?php
+namespace wcf\form;
+
+use wcf\data\user\notification\event\UserNotificationEvent;
+use wcf\data\user\User;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\request\LinkHandler;
+use wcf\system\WCF;
+use wcf\util\HeaderUtil;
+use wcf\util\StringUtil;
+
+/**
+ * Represents the notification unsubscribe form.
+ *
+ * @author	Tim Duesterhus
+ * @copyright	2001-2020 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\Form
+ * @since	5.3
+ */
+class NotificationUnsubscribeForm extends AbstractForm {
+	/**
+	 * user to unsubscribe
+	 * @var	User
+	 */
+	public $user;
+	
+	/**
+	 * user provided verification token
+	 * @var	string
+	 */
+	public $token = '';
+	
+	/**
+	 * whether this is a one-click request
+	 * @var	boolean
+	 */
+	public $isOneClick = false;
+	
+	/**
+	 * notification event to unsubscribe
+	 * @var	UserNotificationEvent
+	 */
+	public $event = null;
+	
+	/**
+	 * Disable security token validation.
+	 */
+	protected function validateSecurityToken() {
+		// Do not validate the security token, the request is authenticated by
+		// the mail token.
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function readParameters() {
+		parent::readParameters();
+		
+		if (isset($_REQUEST['userID'])) {
+			$this->user = new User(intval($_REQUEST['userID']));
+			if (!$this->user->userID) {
+				throw new IllegalLinkException();
+			}
+		}
+		else {
+			throw new IllegalLinkException();
+		}
+		
+		if (isset($_REQUEST['token'])) {
+			$this->token = StringUtil::trim($_REQUEST['token']);
+		}
+		else {
+			throw new IllegalLinkException();
+		}
+		
+		if (!empty($_REQUEST['eventID'])) {
+			$this->event = new UserNotificationEvent(intval($_REQUEST['eventID']));
+			if (!$this->event->eventID) {
+				throw new IllegalLinkException();
+			}
+		}
+		
+		if (!$this->user || !hash_equals($this->user->notificationMailToken, $this->token)) {
```

Comment:
```suggestion
		if (!hash_equals($this->user->notificationMailToken, $this->token)) {
```
`$this->user` cannot be null, see line 60-68.